### DISCUSSION TITLE
Run emscripten inside emscripten? OR Run emscripten native build?

### DISCUSSION THREAD

**Discussion by alienself** [emscripten-core/emscripten#18690] [General]
Hi all!

Is it possible to compile emscripten to wasm and compile c++ from within the browser or from nodejs and produce a wasm file from it? 

I found the emception project: https://github.com/jprendes/emception 
However it is in a broken state, impossible to build as dependencies used have been updated. 
I was wondering if there was an official effort realize such a thing?

Alternatively if this is not possible, would it be possible to somehow bundle emscripten within a native c++ project (like emcc as a static library)? The use case would be to compile c++ code from a single binary without the user having to install the whole toolchain.

**Comment by sbc100** [comment]
I think `emception` is the best effort that we know of.    There is nothing official upstream at this point, but that project clearly proved that it is indeed possible.   Have you considered opening a bug there to see if the author would be interested in updating?

emsdk does a pretty good job of bundling emscripten, but maybe you have some ideas about how to bundle further.   Perhaps you can explain what you mean by "without the user having to install the whole toolchain".. what exactly would that look like?  Which users do you have in mind?

**Reply by alienself** [reply]
Thanks for the fast reply @sbc100!

Agreed, it would be game changing if we had the power to use emscripten directly in the browser. 
I am commenting on open issues to see if someone is able to provide a fix for it. 

As for emsdk, absolutely, the sdk is extremely clear and always a joy to use, however it is possible to use emcc programmatically from C++ and make a native build that would include emscripten, all in a single file (could be dynamically linked too but just as an example)? 

For instance, imagine the use case of a game engine that could provide scripting through C++. 
It would be ideal if the game engine app would embed emscripten directly rather than asking its users to download the SDK on their machine. 

Is anything like this possible?

**Reply by sbc100** [reply]
yes that should be possible.  Indeed I think that is how Unity does it.  They basically just bundle the whole of emskd inside their download.    If you do "emsdk inatall" then then zip up the whole tree that should work when you unzip on the target developers machine.
